File: 240712/최적의 십자 모양 폭발/best-cross-shape-bomb.js
function bomb(arr, n, row, col){
    power = arr[row][col]
    for (let i = 0; i < n; i++){
        for (let j = 0; j < n; j++){
            if ((i > row - power && i < row + power && j === col) || (j > col - power && j < col + power && i === row))
                arr[i][j] = 0
        }
    }

    for (let i = n - 1; i > 0; i--){
        for(let j = 0; j < n; j++){
            if (arr[i][j] == 0){
                arr[i][j] = arr[i - 1][j]
                arr[i - 1][j] = 0
            }
        }
    }
    return arr
}

const fs = require("fs")
const input = fs.readFileSync(0).toString().trim().split("\n")
const n = Number(input[0])
let arr = input.slice(1).map((line) => line.split(" ").map(Number))
let ans = 0
let new_arr

for (let i = 0; i < n; i++){
    for (let j = 0; j < n; j++){
        new_arr = bomb(arr, n, i, j)
        let cnt = 0
        for (let i = 0; i < n; i++){
            for (let j = 0; j < n - 1; j++){
                if (new_arr[i][j] !== 0 && new_arr[i][j] === new_arr[i][j + 1])
                    cnt++
                if (new_arr[j][i] !== 0 && new_arr[j][i] === new_arr[j + 1][i])
                    cnt++
            }
        }
        ans = Math.max(ans, cnt)
        cnt = 0
    }
}

console.log(ans)

File: 240712/최적의 십자 모양 폭발/best-cross-shape-bomb.py
from copy import deepcopy

def bomb(arr, n, row, col):
    new_arr = deepcopy(arr)
    power = int(new_arr[row][col])
    
    for i in range(n):
        for j in range(n):
            if ((i > row - power and i < row + power and j == col) or (j > col - power and j < col + power and i == row)):
                new_arr[i][j] = '0'
    
    for j in range(n):
        tmp = []
        for i in range(n):
            if new_arr[i][j] != '0':
                tmp.append(new_arr[i][j])
        tmp = ['0'] * (n - len(tmp)) + tmp
        for i in range(n):
            new_arr[i][j] = tmp[i]
    
    return new_arr

n = int(input())
arr = [input().split() for _ in range(n)]
ans = 0

for i in range(n):
    for j in range(n):
        new_arr = bomb(arr, n, i, j)
        cnt = 0
        for x in range(n):
            for y in range(n - 1):
                if new_arr[x][y] != '0' and new_arr[x][y] == new_arr[x][y + 1]:
                    cnt += 1
                if new_arr[y][x] != '0' and new_arr[y][x] == new_arr[y + 1][x]:
                    cnt += 1
        ans = max(ans, cnt)

print(ans)